### PR TITLE
Added Neo4j IE

### PR DESCRIPTION
**Pull request by thobalose**
A [Neo4j GIE](https://github.com/SANBI-SA/neo4j_galaxy_ie) based on a modified version of the Neo4j:2.3.3 image  which takes a [neostore datatype](https://github.com/SANBI-SA/tools-sanbi-uwc/tree/master/datatypes/neo4j_datatypes) generated by [this tool](https://github.com/SANBI-SA/tools-sanbi-uwc/tree/master/tools/build_ctb_gene) allowing users to explore a Neo4j Graph database.

### CODE REVIEW

File: config/plugins/interactive_environments/neo/config/neo.ini.sample [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# additional parameters. This command string is re-used for a `docker inspect`
+# command and will likely cause errors if it is extensively modified, past the
+# usual group/sudo changes.
+#command = docker {docker_args}
+
+# The image argument was moved to "allowed_images.yml.sample"
+
+# Additional arguments that are passed to the `docker run` command.
+# NEO4J_UID=1371, NEO4j_GIG=999
+command_inject = --sig-proxy=true -e DEBUG=false -e DEFAULT_CONTAINER_RUNTIME=120 -e NEO4J_UID=none -e NEO4J_GID=none
```
Comment (hexylena) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]: According to https://github.com/SANBI-SA/neo4j_galaxy_ie/blob/master/docker-entrypoint.sh#L7 NEO4J_U/GID are required. Maybe we can default to $UID and $GID?

**Comment by hexylena** [comment]
So excited to see this, but am still thinking over any implementation concerns.

@thobalose did you consider using an official neo4j image as the base? Was there a rationale against doing that?

File: config/plugins/interactive_environments/neo/config/neo.ini.sample (hunk 1 above)
Comment (thobalose) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]: @erasche ,

The tool used to generate the data type [boots up a Neo4j docker container](https://github.com/SANBI-SA/tools-sanbi-uwc/blob/master/tools/build_ctb_gene/build_ctb_gene.py#L110) used load data into a Neo4j database, this runs as the `galaxy` user which is a member of the `docker` group.

The IE then needs to mount the data directory with `rw` access for the user to explore the data.

Comment (hexylena) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]: @thobalose 

> The IE then needs to mount the data directory with rw access for the user to explore the data.

yes. So would it make sense to use `NEO4J_UID=$UID NEO4J_GID=$GID` as the default, in order to have the default command use the same UID and GID as the galaxy user who launches the container and owns the files which have been mounted in the volume?

**Comment by thobalose** [comment]
The official Neo4j docker images exposes, depending on the version, [2-3 ports](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/base/interactive_environments.py#L399) and does not automatically terminate.

Additionally, it assumes/writes as `root`.

**Comment by hexylena** [comment]
@thobalose sure, I meant building on top of that as your FROM image, adding your auto-killing, and port changes.

Pros:
- It might allow your group to more easily import security updates that upstream makes
- you can ignore the other ports that are exposed, since you have some configuration that makes it function with just one port.
- As for reading/writing as root, use you could drop privileges with a `USER` statement, or with a customized `CMD` that runs gosu since you have that installed.

Just thinking about future maintenance and updates.

File: config/plugins/interactive_environments/neo/config/neo.ini.sample (hunk 1 above)
Comment (thobalose) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]: Will have to take another look at this. What do you get when you `$echo $GID` ?

Comment (hexylena) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]:
```
[hxr@leda:~]$ echo $GID
1000
```

when galaxy runs this, it'd be the primary group id of the galaxy user.

Comment (thobalose) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]:
```
thoba at xps in ~
$ echo $GID

thoba at xps in ~
$ echo $GROUPS
1000
```

Comment (thobalose) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]: Added [this](https://github.com/galaxyproject/galaxy/pull/2610/commits/23f8aab6c6f462b4a343101bd95ed46a63005d16).

Comment (hexylena) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample]: Thank you, that is helpful. This way people could at least launch this without changing configuration.

**Comment by thobalose** [comment]
Hi @erasche 

This looks like something Docker is still working on. 
Kindly have a look at [this](https://github.com/docker/docker/issues/3465).

**Comment by hexylena** [comment]
@thobalose yes, but do you actually need to reset them? In the past, I've just ignored the extra volumes, exposed ports with no negative effects.

**Comment by thobalose** [comment]
@erasche 

Do you mean like [so](https://asciinema.org/a/6ki1cpnple4zwue8lr6n7rurm)? 

From there, we can see that ignoring exposed ports adds the built port as an extra one. Thus, having three exposed ports.

File: config/plugins/interactive_environments/neo/config/allowed_images.yml.sample [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# This file lists acceptable images to allow runing.
+#
+# This allows you, the admin, to create multiple flavours
+# for your users to run. E.g. maybe you need a geosciences flavour,
+# you can create the image based on our default image and add the
+# appropriate `apt-get/pip install` statements.
+---
+-
+    image: thoba/neo4j_galaxy_ie
```
Comment (hexylena) [inline config/plugins/interactive_environments/neo/config/allowed_images.yml.sample]: Ok, sorry for one more thing. Would it be possible that you could pin a specific tag, like we do in jupyter? 

https://github.com/galaxyproject/galaxy/blob/dev/config/plugins/interactive_environments/jupyter/config/allowed_images.yml.sample#L9

This is so if/when we change the GIE API, we can help all of the GIE building groups migrate to a newer version of the API with a new tagged image, and cause fewer "Which image are you using?" issues for end users.

**Comment by hexylena** [comment]
@thobalose yes, exactly like that. However that raises a good point, since we started using the `-P` flag in the docker GIE launcher, all ports are exposed rather than a single one being mapped externally (and all other exposed ports not being exposed thanks to `-p`'s behaviour).

Okay, due to the fact that using `-P` will chew up more exposed ports than expected and might decrease number of neo4j GIEs that could be run by heavier deployments, your group has made the right technical decision.

Thank you for working through this with me, I appreciate it!

I had one more last minute comment, otherwise I'm :+1: on this, it should be merged.

@galaxybot test this

**Comment by hexylena** [comment]
ping @martenson can you get galaxy bot to test this?

**Comment by nsoranzo** [comment]
@galaxybot test this

**Pull request merged**